Fix text field selection toolbar under Opacity

liyuqian opened the pull request:
## Description

As we've introduced offset to the Opacity layer, we have to override
`applyTransform` to make Leader/FollowerLayer work correctly.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/30587

Together with https://github.com/flutter/engine/pull/8585,
this test will also exercise test against
https://github.com/flutter/flutter/issues/30586.

## Tests

I added the following tests:
* text field selection toolbar renders correctly inside opacity

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

liyuqian committed:
Check

liyuqian committed:
Fix text field selection toolbar under Opacity

As we've introduced offset to the Opacity layer, we have to override
`applyTransform` to make Leader/FollowerLayer work correctly.

Fixes https://github.com/flutter/flutter/issues/30587

Together with https://github.com/flutter/engine/pull/8585,
this test will also exercise test against
https://github.com/flutter/flutter/issues/30586.

liyuqian commented:
WIP, the test won't pass until
1. We fixed #30587
2. We uploaded the golden file.

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1279,6 +1279,13 @@ class OpacityLayer extends ContainerLayer {
+  void applyTransform(Layer child, Matrix4 transform) {
+    assert(child != null);
+    assert(transform != null);
+    transform.multiply(Matrix4.translationValues(offset.dx, offset.dy, 0.0));
```
there's a Matrix4.translate method, which would make this more efficient

Hixie commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -380,6 +381,58 @@ void main() {
+    await tester.pump();
+
+    // This is needed for the AnimatedOpacity to turn from 0 to 1 so the toolbar is visible.
+    await tester.pumpAndSettle();
```
it's more efficient to just pump once with the required amount of time to skip the animation

liyuqian replied:
I've removed all the previous pumps. However, I couldn't change this `pumpAndSettle` to even `pump(Duration(seconds: 100))` to pass the test. It'll pass with `await tester.pump(); await tester.pump(Duration(seconds: 1));` though. Is this a bug or an expected behavior?

liyuqian replied:
I've also added the first `pump` back in the newest patch. Otherwise, the "COPY/CUT/PASTE" will appear in a different order which causes a golden image mismatch (see attached images). Is this also an expected behavior?
![text_field_opacity_test 0 0](https://user-images.githubusercontent.com/22987568/56320950-30f64280-611a-11e9-9159-7186ea23d862.png)
![text_field_opacity_test 0 0 original](https://user-images.githubusercontent.com/22987568/56320951-32c00600-611a-11e9-9b98-e03aee3ffb87.png)

Hixie replied:
No idea why the order would change...

Hixie replied:
looks like you still have pumpAndSettle?

liyuqian replied:
Yes, I'll either need this `pumpAndSettle`, or `await tester.pump(); await tester.pump(Duration(seconds: 1));` as mentioned in https://github.com/flutter/flutter/pull/31097/files#r276410094. Do you want me to change to `await tester.pump(); await tester.pump(Duration(seconds: 1));`?

Hixie replied:
it's significantly preferred to use pump than pumpAndSettle, because pumpAndSettle will mask regressions.

liyuqian replied:
Changed to `pump`.

dnfield replied:
One suggestion is to capture the return value of `pumpAndSettle` and `expect` the result of it (or expect that it's less than some sane value).

liyuqian replied:
I guess the regression that @Hixie mentioned is about the regression in seconds (e.g., the animation should finish in 1 second but a regression made it 10 seconds). I'm not sure if the return value of `pumpAndSettle` would catch that. What's your thought, @Hixie ?

Hixie replied:
The return value would catch it but it's more likely that someone fixing it will just change the number and it's more likely that a reviewer will fail to notice that.

liyuqian committed:
Matrix.translate and remove unnecessary pump

liyuqian committed:
Add one pump back

Otherwise, the order of "COPY/CUT/PASTE" will be somehow different and
it'll cause a golden image mismatch.

liyuqian commented:
Ping...

Hixie commented on packages/flutter/lib/src/rendering/layer.dart:
```diff
@@ -1279,6 +1279,13 @@ class OpacityLayer extends ContainerLayer {
+    assert(transform != null);
+    transform.translate(offset.dx, offset.dy);
+  }
+
```
fyi @goderbauer

liyuqian committed:
Use pump instead of pumpAndSettle

goderbauer commented on packages/flutter/test/material/text_field_test.dart:
```diff
@@ -380,6 +381,57 @@ void main() {
+      find.byType(Overlay),
+      matchesGoldenFile('text_field_opacity_test.0.0.png'),
+    );
+  }, skip: !Platform.isLinux);
```
Do you really want to skip the entire test on all non-linux platforms? Or just the golden check? In the later, I'd recommend putting the skip on "expectLater". (It looks like everything else should also work on other platforms?)

liyuqian replied:
Good catch! Fixed.

goderbauer reviewed (approved):
LGTM

liyuqian committed:
Only skip golden test

jonahwilliams commented:
This is breaking several devicelab tasks, and failing a golden check. Do these need to be updated, or were they unexpected?  cc @liyuqian
